server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregator.java

```diff
@@ -0,0 +1,269 @@
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(isIpv6, prefixLength, appendPrefixLength, netmask);
+        }
+    }
+
+    final ValuesSource.Bytes valuesSource;
+    final DocValueFormat format;
+    final long minDocCount;
+    final boolean keyed;
+    final BytesKeyedBucketOrds bucketOrds;
+    final IpPrefix ipPrefix;
+
+    public IpPrefixAggregator(
+        String name,
+        AggregatorFactories factories,
+        ValuesSource valuesSource,
+        DocValueFormat format,
+        boolean keyed,
+        long minDocCount,
+        IpPrefix ipPrefix,
+        AggregationContext context,
+        Aggregator parent,
+        CardinalityUpperBound cardinality,
+        Map<String, Object> metadata
+    ) throws IOException {
+        super(name, factories, context, parent, CardinalityUpperBound.MANY, metadata);
+        this.valuesSource = (ValuesSource.Bytes) valuesSource;
+        this.format = format;
+        this.keyed = keyed;
+        this.minDocCount = minDocCount;
+        this.bucketOrds = BytesKeyedBucketOrds.build(bigArrays(), cardinality);
+        this.ipPrefix = ipPrefix;
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        return valuesSource == null
+            ? LeafBucketCollector.NO_OP_COLLECTOR
+            : new IpPrefixLeafCollector(sub, valuesSource.bytesValues(ctx), ipPrefix);
+    }
+
+    private class IpPrefixLeafCollector extends LeafBucketCollectorBase {
+        private final IpPrefix ipPrefix;
+        private final LeafBucketCollector sub;
+        private final SortedBinaryDocValues values;
+
+        IpPrefixLeafCollector(LeafBucketCollector sub, SortedBinaryDocValues values, IpPrefix ipPrefix) {
+            super(sub, values);
+            this.sub = sub;
+            this.values = values;
+            this.ipPrefix = ipPrefix;
+        }
+
+        @Override
+        public void collect(int doc, long owningBucketOrd) throws IOException {
+            if (values.advanceExact(doc)) {
+                int valuesCount = values.docValueCount();
+
+                byte[] previousSubnet = null;
+                for (int i = 0; i < valuesCount; ++i) {
+                    BytesRef value = values.nextValue();
+                    byte[] ipAddress = Arrays.copyOfRange(value.bytes, value.offset, value.offset + value.length);
+                    byte[] netmask = Arrays.copyOfRange(
+                        ipPrefix.netmask.bytes,
+                        ipPrefix.netmask.offset,
+                        ipPrefix.netmask.offset + ipPrefix.netmask.length
+                    );
+                    byte[] subnet = maskIpAddress(ipAddress, netmask);
+                    if (Arrays.equals(subnet, previousSubnet)) {
+                        continue;
+                    }
+                    long bucketOrd = bucketOrds.add(owningBucketOrd, new BytesRef(subnet));
```

Comment:
In the same line as above, we don't need to allocate a new `BytesRef` or its underlying `byte[]` here.  We can create one (or two, since we might need one for holding `previousSubnet`) and reuse them.  That's a pretty common pattern.